### PR TITLE
FEATURE: user status

### PR DESCRIPTION
**Pull request by AndrewPrigorshnev**
This PR adds custom user status to Discourse. The feature is turned off by default. To enable it, set the hidden site setting `enable_user_status`.

Use this button to set custom status:
Screenshot 2022-05-19 at 22 43 35

Then enter status on the opened modal:

Screenshot 2022-05-19 at 22 43 44

For now, it's possible to set only text status, the PR that'll allow adding emoji to status will follow soon. Unless it's implemented, user status will be shown everywhere with the fallback emoji – :mega:.

This PR adds rendering status on the user avatar in the header:

Screenshot 2022-05-19 at 22 51 05

and on the user card:

Screenshot 2022-05-19 at 22 58 13

Also, when status is set, it's shown on the <kbd>Set custom status</kbd> button on the user menu:
Screenshot 2022-05-19 at 22 58 27

### CODE REVIEW

File: app/assets/javascripts/discourse/app/services/user-status.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    this.appEvents.trigger("do-not-disturb:changed");
+  }
+
+  async clear() {
+    await ajax({
+      url: "/user-status.json",
+      type: "DELETE",
+    });
+
+    this.currentUser.status = null;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/user-status.js:24]: Shouldn't this use a set?

File: app/assets/javascripts/discourse/app/templates/modal/user-status.hbs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{{#d-modal-body}}
+  {{#conditional-loading-spinner condition=loading}}
+    <div class="control-group user-status-description-wrap">
+      {{input
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/templates/modal/user-status.hbs:4]: We don't enforce it yet, but ftr correct indentation is:

```
{{input
  foo=1
  bar="baz"
}}
```

File: app/controllers/user_status_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  def clear
+    ensure_feature_enabled
+    current_user.clear_status
+    render json: success_json
+  end
+
+  private
+
+  def ensure_feature_enabled
+    raise ActionController::RoutingError.new("Not Found") if !SiteSetting.enable_user_status
```
Comment (jjaffeux) [inline app/controllers/user_status_controller.rb:23]: I don't know what we do at other places, but shouldn't this be a 405 in this case?

File: app/assets/stylesheets/common/base/discourse.scss [hunk 4]
```diff
@@ -424,6 +424,29 @@ table {
...
+.user-status-background {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  width: 1.25em;
+  height: 1.25em;
+  background-color: var(--tertiary-low);
+  border-radius: 50%;
+
+  img.emoji {
```
Comment (jjaffeux) [inline app/assets/stylesheets/common/base/discourse.scss:443]: is the img necessary here?

File: app/assets/javascripts/discourse/app/services/user-status.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  @service appEvents;
+
+  async set(status) {
+    await ajax({
+      url: "/user-status.json",
+      type: "PUT",
+      data: { description: status.description },
+    });
+
+    this.currentUser.status = status;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/user-status.js:14]: Same it seems like it should use a set

File: app/serializers/user_card_serializer.rb [hunk 6]
```diff
@@ -222,6 +223,10 @@ def card_background_upload_url
     object.card_background_upload&.url
   end
 
+  def status
```
Comment (jjaffeux) [inline app/serializers/user_card_serializer.rb:226]: Anything we need to consider here https://github.com/discourse/discourse/blob/main/app/controllers/users_controller.rb#L108 ?

File: app/assets/javascripts/discourse/app/services/user-status.js (hunk 1 above)
Comment (AndrewPrigorshnev) [reply app/assets/javascripts/discourse/app/services/user-status.js]: We should, thanks!

File: app/assets/javascripts/discourse/app/templates/modal/user-status.hbs (hunk 2 above)
Comment (AndrewPrigorshnev) [reply app/assets/javascripts/discourse/app/templates/modal/user-status.hbs:4]: ☑️

File: app/controllers/user_status_controller.rb (hunk 3 above)
Comment (AndrewPrigorshnev) [reply app/controllers/user_status_controller.rb]: I haven't found places where we return 405, but I agree that will be much better, I've changed it.

File: app/assets/stylesheets/common/base/discourse.scss (hunk 4 above)
Comment (AndrewPrigorshnev) [reply app/assets/stylesheets/common/base/discourse.scss]: We'd better go without `img`.

File: app/models/user.rb [hunk 7]
```diff
@@ -1500,6 +1501,23 @@ def display_name
...
   end
 
+  def clear_status
+    user_status.destroy if user_status
+  end
+
+  def set_status(description)
+    now = Time.zone.now
+    if user_status
+      user_status.update(description: description, set_at: now)
```
Comment (jjaffeux) [inline app/models/user.rb:1511]: Maybe it should be `update!` ?

File: app/serializers/user_card_serializer.rb (hunk 6 above)
Comment (AndrewPrigorshnev) [reply app/serializers/user_card_serializer.rb:226]: This method returns status without additional changes (I've just tested it). I wonder if you know where we use this method? Quick search on the client-side do not return any usage.

Comment (jjaffeux) [reply app/serializers/user_card_serializer.rb:226]: I was more concerned about maybe having to preload something to query in the serializer

Comment (jjaffeux) [reply app/serializers/user_card_serializer.rb:226]: Also you are right I don't know where user-cards or cards is used 🤔  I can see @tgxworld made specific changes and tests in discourse-follow to not break this route, maybe he knows?

Comment (AndrewPrigorshnev) [reply app/serializers/user_card_serializer.rb:226]: Anyway, preloading won't hurt here, so I've added it (still curious where we use it, hope tgxworld know)

Comment (AndrewPrigorshnev) [reply app/serializers/user_card_serializer.rb:226]: It was introduced in https://github.com/discourse/discourse/pull/9078 so maybe @davidtaylorhq remember something.

Comment (davidtaylorhq) [reply app/serializers/user_card_serializer.rb:226]: The multi-user-card route was added for https://meta.discourse.org/t/user-card-directory/144479?u=david (Internal ref https://dev.discourse.org/t/24089)

**Comment by jjaffeux** [comment]
> Screenshot 2022-05-19 at 22 43 44

Nitpicking here, but can we align trash button to right side?

**Comment by jjaffeux** [comment]
Also do we need to show trash when you didn't have any status yet?

File: app/serializers/user_card_serializer.rb (hunk 6 above)
Comment (jjaffeux) [reply app/serializers/user_card_serializer.rb:226]: So a core route only used in one theme, well 🤷‍♂️  At least we have the answer :p thanks

**Comment by AndrewPrigorshnev** [comment]
> Also do we need to show trash when you didn't have any status yet?

Yes, we should hide it in this case. This is a nice catch, somehow I forgot about it and left this job halfway done. Now it's fixed.

File: app/assets/javascripts/discourse/app/controllers/user-status.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import { action } from "@ember/object";
+import { notEmpty } from "@ember/object/computed";
+import { inject as service } from "@ember/service";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+import bootbox from "bootbox";
+
+export default Controller.extend(ModalFunctionality, {
+  userStatusService: service("user-status"),
+
+  description: "",
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/controllers/user-status.js:12]: Im usually not a big fan of empty strings, is there a reason to prefer this over `null` ?

File: app/assets/javascripts/discourse/app/templates/components/user-card-contents.hbs [hunk 9]
```diff
@@ -62,6 +62,9 @@
           {{#if this.user.staged}}
             <h2 class="staged">{{i18n "user.staged"}}</h2>
           {{/if}}
+          {{#if siteSettings.enable_user_status}}
+            <h3 class="user-status">{{html-safe this.userStatus}}</h3>
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/templates/components/user-card-contents.hbs:66]: You have a security issue here, nothing stops a user to have malicious code in `userStatus`, I could for example have this css as my status:

```
<style>body {display: none;}</style>
```

And anyone looking at my user card would have a blank screen, there are most likely way more dangerous attacks than this possible :)

File: config/site_settings.yml [hunk 10]
```diff
@@ -361,6 +361,10 @@ basic:
     default: true
   sitemap_page_size:
     default: 10000
+  enable_user_status:
+    hidden: true
```
Comment (jjaffeux) [inline config/site_settings.yml:365]: I assume the hidden is on purpose for now?

**Review by jjaffeux — CHANGES_REQUESTED** [review]
Security issue needs to be handled

File: app/assets/javascripts/discourse/app/controllers/user-status.js (hunk 8 above)
Comment (tgxworld) [reply app/assets/javascripts/discourse/app/controllers/user-status.js:12]: The usual argument is that it is usually easier to deal with a single data type `String` instead of having to deal with two different data types.

File: app/controllers/user_status_controller.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  def clear
+    ensure_feature_enabled
+    current_user.clear_status
+    render json: success_json
+  end
+
+  private
+
+  def ensure_feature_enabled
+    raise ActionController::MethodNotAllowed.new if !SiteSetting.enable_user_status
```
Comment (tgxworld) [inline app/controllers/user_status_controller.rb:23]: I'm curious what the reason is for choosing `MethodNotAllowed` which returns a `405`. I kind of think `403` is the more correct response here when the site setting is disabled since the method is actually allowed but the server is choosing not to fulfill the request not because of the request method.

File: app/models/user.rb [hunk 12]
```diff
@@ -1500,6 +1501,23 @@ def display_name
     end
   end
 
+  def clear_status
+    user_status.destroy if user_status
```
Comment (tgxworld) [inline app/models/user.rb:1505]:
```suggestion
    user_status.destroy! if user_status
```

File: app/models/user_status.rb [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class UserStatus < ActiveRecord::Base
+  belongs_to :user
+
+  validate :ends_at_greater_than_set_at
```
Comment (tgxworld) [inline app/models/user_status.rb:6]: I think we will only need to run this validation when `ends_at` and `set_at` are going to change.

File: config/locales/server.en.yml [hunk 14]
```diff
@@ -5205,3 +5205,7 @@ en:
     small_action_post_raw: "Continue discussion at %{new_title}."
 
   fallback_username: "user"
+
+  user_status:
+    errors:
+      ends_at_should_be_greater_than_set_at: "ends_at should be greater than set at"
```
Comment (tgxworld) [inline config/locales/server.en.yml:5211]:
```suggestion
      ends_at_should_be_greater_than_set_at: "ends_at should be greater than set_at"
```

File: app/serializers/user_card_serializer.rb (hunk 6 above)
Comment (tgxworld) [inline app/serializers/user_card_serializer.rb:226]: I think what we usually prefer to do is to define a `include_status?` method instead. Otherwise, we're going to be sending down `status: null` when the site setting is disabled.

File: app/models/user.rb [hunk 15]
```diff
@@ -1500,6 +1501,23 @@ def display_name
...
+  end
+
+  def set_status(description)
```
Comment (tgxworld) [inline app/models/user.rb:1508]:
```suggestion
  def set_status!(description)
```

Minor suggestion to reflect that this method raises an exception if it fails

File: app/serializers/current_user_serializer.rb [hunk 16]
```diff
@@ -336,4 +337,8 @@ def experimental_sidebar_enabled
   def include_experimental_sidebar_enabled?
     SiteSetting.enable_experimental_sidebar
   end
+
+  def status
```
Comment (tgxworld) [inline app/serializers/current_user_serializer.rb:341]: Same concern as https://github.com/discourse/discourse/pull/16875/files#r878953293https://github.com/discourse/discourse/pull/16875/files#r878953293

File: app/assets/javascripts/discourse/app/controllers/user-status.js (hunk 8 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/controllers/user-status.js:12]: Yep agreed on backend, but I don't think that works well in frontend code. As this example shows you end up having a lot of `if (foo.length)` also as this is not type you are never sure if it's `null` or `""` so you actually use something like `foo?.length` or any `not empty` kind method.

File: app/controllers/user_status_controller.rb (hunk 11 above)
Comment (jjaffeux) [reply app/controllers/user_status_controller.rb:23]: Yes 403 would work for me too, I was mostly annoyed by 404

File: app/assets/javascripts/discourse/app/controllers/user-status.js (hunk 8 above)
Comment (AndrewPrigorshnev) [reply app/assets/javascripts/discourse/app/controllers/user-status.js]: I think I more agree with @jjaffeux here. I switched it to using `null`, it makes code simpler. 

But soon we may decide to switch it back to an empty string 😄. The reason is that when we add emoji picker to the status modal, we're going to allow users to set status that contains only emoji. It may turn out that an empty string will  be more convenient at that moment. But, we'll see when implementing that.

File: app/assets/javascripts/discourse/app/templates/components/user-card-contents.hbs (hunk 9 above)
Comment (AndrewPrigorshnev) [reply app/assets/javascripts/discourse/app/templates/components/user-card-contents.hbs]: That was a huge flaw! Thank you for catching this and for the explanation!

File: config/site_settings.yml (hunk 10 above)
Comment (AndrewPrigorshnev) [reply config/site_settings.yml:365]: Yep, this is the very first iteration, we're going to add more functionality to it soon.

File: app/models/user.rb (hunk 12 above)
Comment (AndrewPrigorshnev) [reply app/models/user.rb]: A good catch! I've fixed it.

File: app/serializers/current_user_serializer.rb (hunk 16 above)
Comment (AndrewPrigorshnev) [reply app/serializers/current_user_serializer.rb:345]: This also fixed.

File: app/models/user_status.rb (hunk 13 above)
Comment (AndrewPrigorshnev) [reply app/models/user_status.rb:6]: I'm not sure here. Do you mean that it's better to remove the validator for now and then return it back to this class when we'll add code that will be setting `ends_at`? Or do you mean that this shouldn't be implemented as a validator on the model, and we need to check it at the place where we set `ends_at`?

File: app/controllers/user_status_controller.rb (hunk 11 above)
Comment (jjaffeux) [reply app/controllers/user_status_controller.rb:23]: No strong dislike :) These semantics are always a little bit hard to follow, I would vote for consistency we core, if we usually go with 403 for this, go with 403, if 404 go for 404...

Comment (tgxworld) [reply app/controllers/user_status_controller.rb:23]: Searching through core, it seems like we return a 404 as a convention.

File: app/models/user_status.rb (hunk 13 above)
Comment (tgxworld) [reply app/models/user_status.rb:6]: O sorry I mean that this validation will only need to run when `will_save_change_to_ends_at` or `will_save_change_to_set_at` is `true`.

Comment (AndrewPrigorshnev) [reply app/models/user_status.rb]: Ok, got it. I've added a condition to the validator.

File: app/controllers/user_status_controller.rb (hunk 11 above)
Comment (AndrewPrigorshnev) [reply app/controllers/user_status_controller.rb]: Ok, we go with 404.

**Comment by AndrewPrigorshnev** [comment]
I've addressed all the comments.

**Review by jjaffeux — APPROVED** [review]
